ejb3/src/main/java/org/jboss/as/ejb3/component/stateful/StatefulSessionSynchronizationInterceptor.java

```diff
@@ -132,27 +132,38 @@ public Object processInvocation(final InterceptorContext context) throws Excepti
                     }
                 }
                 // proceed with the invocation
+                // handle exceptions to coincide with exception handling in StatefulComponentInstanceInterceptor
                 try {
                     return context.proceed();
-                } catch (Exception e) {
-                    if(component.shouldDiscard(e, context.getMethod())) {
+                } catch (Exception ex) {
+                    if(component.shouldDiscard(ex, context.getMethod())) {
                         toDiscard = true;
                     }
+                    throw ex;
+                } catch (Error e) {
```

Comment:
Since exception handling is the same for both Error and Throwable, you should combine these into a multi-catch:
catch (Error | Throwable e) { ... }